### PR TITLE
Rust hanoi

### PR DESCRIPTION
**Pull request by First529**

### CODE REVIEW

File: src/bin/traditional.rs [hunk 1]
```diff
@@ -45,6 +49,85 @@ fn hi(_: PotatoRequest) -> PotatoResponse {
         .add_header("Content-Length", &body.len().to_string())
 }
 
+fn simple_add(req: PotatoRequest) -> PotatoResponse {
+    let res = PotatoResponse::new();
+
+    let num_str = req.body.unwrap();
```
Comment (wisaTong) [inline src/bin/traditional.rs]: Don't unwrap body. Handle the case where there is no body provided and return 400 Bad request.

File: src/bin/traditional.rs [hunk 2]
```diff
@@ -45,6 +49,85 @@ fn hi(_: PotatoRequest) -> PotatoResponse {
...
+    let num_str = str::from_utf8(&num_str).unwrap();
```
Comment (wisaTong) [inline src/bin/traditional.rs]:
```suggestion
    let num_str = String::from_utf8_lossy(&num_str);
```
Use String's lossy so there is no unwrap

File: src/bin/traditional.rs [hunk 3]
```diff
@@ -45,6 +49,85 @@ fn hi(_: PotatoRequest) -> PotatoResponse {
...
+        compute_hanoi(num - 1, via, to, from);
+    }
+}
+
+fn hanoi(req: PotatoRequest) -> PotatoResponse {
+    let res = PotatoResponse::new();
+
+    let num_str = req.body.unwrap();
+    let num_str = str::from_utf8(&num_str).unwrap();
+    println!("{}", num_str);
```
Comment (wisaTong) [inline src/bin/traditional.rs]: Don't print

File: src/bin/traditional.rs [hunk 4]
```diff
@@ -45,6 +49,85 @@ fn hi(_: PotatoRequest) -> PotatoResponse {
...
+    compute_hanoi(num_str.parse::<i32>().unwrap() , 1, 2, 3);
+
+    let result = "Sucess";
```
Comment (wisaTong) [inline src/bin/traditional.rs]: Typo

File: src/request.rs [hunk 5]
```diff
@@ -42,16 +57,49 @@ pub struct PotatoRequest {
...
+                continue;
+            }
+            let (a, b) = i.split_at(i.find(":").unwrap());
+            if a.trim().starts_with("Content-Length") {
+                content_len = b.replace(": ", "").to_string().parse::<usize>().unwrap();
+            }
         }
+        let body = raw[bindex + 4..content_len + bindex + 4].to_vec();
+
+        PotatoRequest::new(HttpRequestMethod::GET, &headers, Some(body))
```
Comment (wisaTong) [inline src/request.rs]: Why GET?

File: src/server.rs [hunk 6]
```diff
@@ -113,4 +116,32 @@ impl<'a> PotatoServer<'a> {
...
+                continue;
+            }
+            if x.starts_with("-") {
+                check = false;
+            }
+            if check {
+                if x.is_empty() {
+                    break;
+                }
+                println!("{}", x);
```
Comment (wisaTong) [inline src/server.rs]: Dont print

File: src/server.rs [hunk 7]
```diff
@@ -113,4 +116,32 @@ impl<'a> PotatoServer<'a> {
         stream.write(&res).unwrap();
         stream.flush().unwrap();
     }
+
+    fn body_check(&self, checker: bool, line: &str) {
```
Comment (wisaTong) [inline src/server.rs]: What is this function

File: src/server.rs [hunk 8]
```diff
@@ -113,4 +116,32 @@ impl<'a> PotatoServer<'a> {
...
+        if checker {
+            println!("{}", line);
+        }
+    }
+
+    pub fn get_header(&self, s: &str, ignore: &str) -> HashMap<String, String> {
+        let mut check: bool = true;
+        let mut tmp: HashMap<String, String> = HashMap::new();
```
Comment (wisaTong) [inline src/server.rs]: `tmp` -> `headers`

**Review by wisaTong — CHANGES_REQUESTED** [review]

**Comment by wisaTong** [comment]
@booknap please fix your commit email

**Review by wisaTong — APPROVED** [review]

File: src/server.rs (hunk 7 above)
Comment (Vichacorn) [reply src/server.rs]: unused fn. already remove it.

File: src/request.rs (hunk 5 above)
Comment (Vichacorn) [reply src/request.rs]: Change to OPTION(any method)

Comment (wisaTong) [reply src/request.rs]: You have `head` variable above which should looks something like `[METHOD] [PATH] HTTP/1.1`. You need to extract `METHOD` and `PATH` and set it in `PotatoRequest`

**Comment by wisaTong** [comment]
LGTM Merge ley

**Pull request merged**